sched/scheduler/cluster_state.go

```diff
@@ -137,62 +146,95 @@ func (c *clusterState) updateCluster() {
+					ns.timeFlaky = nilTime
+					c.nodes[update.Id] = ns
+					delete(c.suspendedNodes, update.Id)
+					log.Infof("Recovered suspended node %v (%#v), now have %d healthy nodes (%d suspended)",
+						update.Id, ns, len(c.nodes), len(c.suspendedNodes))
+				}
+
 			} else if ns, ok := c.nodes[update.Id]; !ok {
-				// This is a new unrecognized node, add it to the cluster.
-				c.nodes[update.Id] = newNodeState(update.Node)
+				// This is a new unrecognized node, add it to the cluster, possibly in a suspended state.
+				if c.readyFn == nil || c.readyFn(ns.node) {
+					// We're either not checking readiness or it is ready, skip suspended state and add this as a healthy node.
+					c.nodes[update.Id] = newNodeState(update.Node)
+					log.Infof("Added new node: %v (%#v), now have %d healthy nodes (%d suspended)",
+						update.Id, update.Node, len(c.nodes), len(c.suspendedNodes))
+				} else {
+					// Add this to the suspended nodes to be checked on every update().
+					c.suspendedNodes[update.Id] = newNodeState(update.Node)
+					c.suspendedNodes[update.Id].timeInit = time.Now()
+					log.Infof("Added new suspended node: %v (%#v), now have %d healthy nodes (%d suspended)",
+						update.Id, update.Node, len(c.nodes), len(c.suspendedNodes))
+				}
 				c.nodeGroups[""].idle[update.Id] = c.nodes[update.Id]
-				log.Infof("Added new node: %v (%+v), now have %d nodes", update.Id, update.Node, len(c.nodes))
+
 			} else {
 				// This node is already present, log this spurious add.
-				log.Infof("Node already added!! %v (%v)", update.Id, ns)
+				log.Infof("Node already added!! %v (%#v)", update.Id, ns)
 			}
 		case cluster.NodeRemoved:
 			if ns, ok := c.suspendedNodes[update.Id]; ok {
-				// Node already suspended, make sure it's now marked as lost and not flaky.
-				log.Infof("Node already marked suspended (was flaky=%t) : %v (%v)", ns.timeFlaky != nilTime, update.Id, ns)
+				// Node already suspended, make sure it's now marked as lost and not flaky (keep readiness status intact).
+				log.Infof("Already suspended node marked as removed: %v (was %#v)", update.Id, ns)
 				ns.timeLost = time.Now()
 				ns.timeFlaky = nilTime
+
 			} else if ns, ok := c.nodes[update.Id]; ok {
 				// This was a healthy node, mark it as lost now.
 				ns.timeLost = time.Now()
 				c.suspendedNodes[update.Id] = ns
 				delete(c.nodes, update.Id)
-				log.Infof("Removing node by marking as lost: %v (%v), now have %d nodes, %d suspended",
+				log.Infof("Removing node by marking as lost: %v (%#v), now have %d nodes (%d suspended)",
 					update.Id, ns, len(c.nodes), len(c.suspendedNodes))
+
 			} else {
 				// We don't know about this node, log spurious remove.
 				log.Infof("Cannot remove unknown node: %v", update.Id)
 			}
 		}
 	}
 
-	// Clean up lost nodes that haven't recovered in time, and put flaky nodes back in rotation.
+	// Clean up lost nodes that haven't recovered in time, add flaky nodes back into rotation after some time,
```

Comment:
I think this allows nodes that are stuck initializing to be stuck forever...also not sure why we are reinstating flaky based on time... alternative approach 

Have an other loop (in scheduler?)  that is checking all suspendedNodes for good response to 'StatusAll', if get healthy status, reinstate the node, at end of loop drop all nodes that have been suspended too long.